apis/v1alpha1/gateway_types.go

```diff
@@ -249,6 +249,39 @@ type Listener struct {
 	// is required if the Protocol field is "HTTPS" or "TLS" and
 	// ignored otherwise.
 	//
+	// The association of SNIs to Certificate defined in TLSConfig is
+	// defined based on the value of HostnameMatchType for this listener:
+	// - "Domain": Certificate should be used used for the domain and its
+	//   first-level subdomains.
+	// - "Exact": Certificate should be used for the domain only.
+	// - "Any": Certificate in TLSConfig is the default certificate to use.
+	//
+	// The GatewayClass must use the most specific available certificate.
+	// For example, consider the following two compatible listeners:
+	// - hostname:
+	//     name: example.com
+	//     match: Domain
+	//   port: 443
+	//   protocol: HTTPS
+	//   tls:
+	//     certificateRef:
+	//       resource: secrets
+	//       group: core
+	//       name: wildcard-cert
+	// - hostname:
+	//     name: foo.example.com
+	//     match: Exact
+	//   port: 443
+	//   protocol: HTTPS
+	//   tls:
+	//     certificateRef:
+	//       resource: secrets
+	//       group: core
+	//       name: foo-example-com-cert
```

Comment:
I like this example but am scared that it could easily become stale as we change the API. Maybe we can simply say the certificate associated with the longest matching hostname should be chosen here?

Reply:
Fair point. I'll remove this and summarize in text in the next rebase.